File: patterns/stack/largest-rectangle/largest-rectangle.py
# Leetcode 84: Largest Rectangle in Histogram
# Leetcode 84: Largest Rectangle in Histogram
# Pattern: Stack
# Stack Type: Monotonic

from typing import List

def largestRectangleArea(heights: List[int]) -> int:
    maxArea = 0
    stack = []
    
    heights.append(0)
    
    for i in range(len(heights)):
        while len(stack) > 0 and heights[i] < heights[stack[- 1]]:
            height = heights[stack.pop()]
            width = i if not stack else i - stack[-1] - 1

            maxArea = max(maxArea, height * width)
        
        stack.append(i)
        
    return maxArea

print(largestRectangleArea([2,1,5,6,2,3]))
print(largestRectangleArea([2,4]))

File: patterns/stack/largest-rectangle/largest-rectangle.ts
/**
 * Leetcode 84: Daily Largest Rectangle in Histogram
 * https://leetcode.com/problems/largest-merge-of-two-strings/
 * Pattern: Stack
 * Stack Type: Monotonic
 * Language: Typescript
 * Insight: Last unresolved element decides next action
 */
function largestRectangleArea(heights: number[]): number {
    const stack: number[] =  [];
    let maxArea = 0;
    const n = heights.length;

    heights.push(0)

    for (let i = 0; i < n; i++) {
        while(stack.length > 0 && heights[i] > heights[stack[stack.length - 1]]) {
            const height = heights[stack.pop()!];

            const width = stack.length === 0 ? i : i - stack[stack.length - 1] - 1;
            maxArea = Math.max(maxArea, height * width)
        }

        stack.push(i)
    }
    return maxArea
};

